components/scream/src/physics/spa/tests/spa_read_data_from_file_test.cpp

```diff
@@ -13,156 +12,145 @@
-  {
-    Real ccn3 = 0.0;
-    for (int i=1;i<=ncols;i++) {
-      Real wgt = 1.0 / std::pow(2.0,i);
-      if (i == ncols) {wgt *= 2.0;}
-      ccn3 += wgt * (klev*1.0 + t*10.0 + i*100.0);
-    }
-    return ccn3;
-  } // ccn3_func
-  //
-  static Real aer_func(const Int t, const Int icol, const Int bnd, const Int klev, const Int ncols, const Int mode)
-  {
-    Real aer_out = 0.0;
-    for (int i=1;i<=ncols;i++) {
-      Real wgt = 1.0 / std::pow(2.0,i);
-      if (i == ncols) {wgt *= 2.0;}
-      if (mode==0) {  // G
-        aer_out += wgt * t;
-      } else if (mode==1) { // SSA
-        aer_out += wgt * (i-1);
-      } else if (mode==2) { // TAU
-        aer_out += wgt * bnd;
-      } else if (mode==3) { // LW-TAU
-        aer_out += wgt * klev;
+template <typename S>
+using view_1d = typename KokkosTypes<DefaultDevice>::template view_1d<S>;
+
+Real ps_func(const Int t, const Int icol, const Int ncols);
+Real ccn3_func(const Int t, const Int icol, const Int klev, const Int ncols);
+Real aer_func(const Int t, const Int icol, const Int bnd, const Int klev, const Int ncols, const Int mode);
+
+TEST_CASE("spa_read_data","spa")
+{
+  // Set up the mpi communicator and init the pio subsystem
+  ekat::Comm spa_comm(MPI_COMM_WORLD);  // MPI communicator group used for I/O set as ekat object.
+  MPI_Fint fcomm = MPI_Comm_c2f(spa_comm.mpi_comm());  // MPI communicator group used for I/O.  In our simple test we use MPI_COMM_WORLD, however a subset could be used.
+  scorpio::eam_init_pio_subsystem(fcomm);   // Gather the initial PIO subsystem data creater by component coupler
+
+  // Establish the SPA function object
+  using SPAFunc = spa::SPAFunctions<Real, DefaultDevice>;
+  using Spack = SPAFunc::Spack;
+
+  std::string spa_data_file = "spa_data_for_testing.nc";
+  std::string spa_remap_file = "spa_data_for_testing.nc";
+  Int max_time = 3;
+  Int ncols    = 48;
+  Int nlevs    = 4;
+  Int nswbands = 2;
+  Int nlwbands = 3;
+  // Break the test set of columns into local degrees of freedom per mpi rank
+  auto comm_size = spa_comm.size();
+  auto comm_rank = spa_comm.rank();
+  std::vector<int> my_dofs;
+  for (int ii=comm_rank;ii<ncols;ii+=comm_size) {
+    my_dofs.push_back(ii);
+  }
+  view_1d<int> dofs_gids("",my_dofs.size());
+  auto dofs_gids_h = Kokkos::create_mirror_view(dofs_gids);
+  Kokkos::deep_copy(dofs_gids_h,dofs_gids);
+  for (int ii=0;ii<my_dofs.size();ii++) {
+    dofs_gids_h(ii) = my_dofs[ii];
+  }
+  Kokkos::deep_copy(dofs_gids,dofs_gids_h);
+  // Set up the set of SPA structures needed to run the test
+  SPAFunc::SPAHorizInterp spa_horiz_interp;
+  SPAFunc::get_remap_weights_from_file(spa_remap_file,ncols,dofs_gids,spa_horiz_interp);
+  SPAFunc::SPAData spa_data(dofs_gids.size(), nlevs, nswbands, nlwbands);
+
+  // Verify that the interpolated values match the algorithm for the data and the weights.
+  //       weights(i) = 1 / (2**i), weights(-1) = 1 / (2**(i-1)) such that sum(weights) = 1., for i=0,1,2
```

Comment:
Nitpicky: the formula for `weights(-1)` depends on i, but I assume it shouldn't?

Reply:
good catch. you are correct, it is dependent on the number of cols